### ISSUE TITLE
[Usage]: How to get the router logits for an MoE model?

### ISSUE THREAD

**Issue by zhenqincn** [vllm-project/vllm#19342]
### Your current environment

I would like to run a Qwen/Qwen3-30B-A3B model and need to store the router logits for each prompt in order to calculate the usage balance among experts. Is there a recommended or efficient way to implement this?

### How would you like to use vllm

I want to run inference of a [specific model](put link here). I don't know how to integrate it with vllm.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by ZhaozwTD** [comment]
Is there an effective method?

**Comment by mohsenfayyaz** [comment]
Not a direct solution within vLLM itself, but in my [recent project](https://github.com/adobe-research/SteerMoE) I modified the modeling code to save the router logits to a file. You can see an example here:
https://github.com/adobe-research/SteerMoE/blob/main/src/modeling_vllm_save/qwen3_moe.py

A few considerations:
- You’ll need to make the same modification in each modeling file you want to track.
- It’s easier to run with batch_size=1.
- Set `enable_prefix_caching=False` to avoid skipping tokens.
- You’ll also need to handle reading the saved logits after each generation externally, as shown [here](https://github.com/adobe-research/SteerMoE/blob/main/custom_steering.ipynb)

**Comment by ZhaozwTD** [comment]
many thanks~

> Not a direct solution within vLLM itself, but in my [recent project](https://github.com/adobe-research/SteerMoE) I modified the modeling code to save the router logits to a file. You can see an example here: https://github.com/adobe-research/SteerMoE/blob/main/src/modeling_vllm_save/qwen3_moe.py
> 
> A few considerations:
> 
> * You’ll need to make the same modification in each modeling file you want to track.
> * It’s easier to run with batch_size=1.
> * Set `enable_prefix_caching=False` to avoid skipping tokens.
> * You’ll also need to handle reading the saved logits after each generation externally, as shown [here](https://github.com/adobe-research/SteerMoE/blob/main/custom_steering.ipynb)
